### PR TITLE
Add Pyodide support and CI jobs for `statsmodels`

### LINKED ISSUE #9166: Feature request: out-of-tree Pyodide builds for `statsmodels`
#### Is your feature request related to a problem? Please describe

Hi there! I am opening this feature request to gauge ideas and comments about out-of-tree Pyodide builds, i.e., wasm32 wheels via [the Emscripten toolchain](https://emscripten.org/) for `statsmodels`. In my most recent work assignment, I am working on improving the interoperability for the Scientific Python ecosystem of packages with Pyodide and with each other, which shall culminate with efforts towards bringing interactive documentation for these packages where they can then be run in [JupyterLite notebooks](https://jupyter.org/try-jupyter/lab/index.html), through nightly builds and wheels for these packages pushed to PyPI-like indices on Anaconda, at and during a later phase during the project.

It looks like in-tree builds in the Pyodide have been built in the past as noted by the conversations in https://github.com/statsmodels/statsmodels/issues/7956 –  and they seem to be maintained with every release for Pyodide. However, this issue proposes out-of-tree builds for `statsmodels` on its own CI and build infrastructure. I would be glad to work on this for `statsmodels`.

#### Describe the solution you'd like

1. A CI pipeline on Azure Pipelines or GitHub Actions (I would prefer the latter, but I have no qualms with the former) where Emscripten/Pyodide builds for the development version of `statsmodels` are pursued
2. Testing the built wheels against a Pyodide wasm32 runtime virtual environment within the same workflow
3. Fixing up and skipping failing tests as necessary based on current Pyodide limitations and ensuring that all relevant test cases pass

#### Describe alternatives you have considered

N/A

#### Additional context

This project is being tracked at https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/18 and has started out with packages like PyWavelets (https://github.com/PyWavelets/pywt/pull/701) and NumPy (https://github.com/numpy/numpy/pull/25894) being complete, thanks to @rgommers. Other packages besides `statsmodels`, such as `matplotlib`, `zarr`, `pandas`, and many more are planned to follow suit soon in the coming months.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
- [x] closes #9166
- [x] tests added/passed. 
- [ ] code/documentation is well formatted.  
- [ ] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message).

This PR adds a GitHub Actions CI job in `emscripten.yml` that builds and tests WASM wheels for `statsmodels` (Pyodide version 0.26.0, Python 3.12), compiled using the Emscripten toolchain (version 3.1.58). The `pytest` test suite is then executed against the built wheel using the in-house `statsmodels.test()` API – the tests have been fixed or skipped in order to adapt to the limitations of Pyodide in the browser, such as not being able to use `threading`/`multiprocessing` from the standard library or the fact that the bitness for the Python interpreter that is powered by Pyodide is 32-bit.

A Node.js-based test suite that calls Pyodide from JS is used instead of the Pyodide-managed special virtual environment due to differences in being able to load shared objects, where there are some problems for `statsmodels` regarding the visibility of symbols most likely coming from OpenBLAS, please see previous discussion(s) on #9166 for more information.

### Additional context

Here are a few xrefs to some other PRs that have been implemented for packages in the Scientific Python ecosystem:

- NumPy's Pyodide job: https://github.com/numpy/numpy/pull/25894, by courtesy of @rgommers
- The same for PyWavelets (an optional dependency for `scikit-image`): PyWavelets/pywt#701 and PyWavelets/pywt#714
- For the `pandas` repository: pandas-dev/pandas#57896
- scikit-image/scikit-image#7350
- and so on

The greater, long-term goal towards this change is to implement Sphinx-based interactive documentation via JupyterLite and WASM-powered, in-browser kernels, as referenced in Quansight-Labs/czi-scientific-python-mgmt#19, see also: Quansight-Labs/czi-scientific-python-mgmt#18. A pilot that can be readily tried out is available for the "API reference" pages under the [PyWavelets documentation](https://pywavelets.readthedocs.io/en/latest/ref/wavelets.html). This will be preceded by configuring a place to publish these WebAssembly/Emscripten wheels nightly or on a scheduled cadence for `statsmodels` (which could be a third-party, PyPI-like index such as Anaconda.org) and then integrating Sphinx extensions such as [`jupyterlite-sphinx`](https://github.com/jupyterlite/jupyterlite-sphinx) for hosted documentation. Work towards these areas is going on in-line with the above tracking issue(s).

<hr>

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify your changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it helps 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

File: tools/ci/run_statsmodels_test_suite.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        `);
+
+
+        await pyodide.runPythonAsync("import micropip; micropip.install('pytest')");
+        await pyodide.runPythonAsync("import micropip; micropip.install('pytest-cov')");
+        await pyodide.runPythonAsync("import micropip; micropip.install('matplotlib')");
+        await pyodide.runPythonAsync(`
+        import sys
+        import statsmodels
+        res = statsmodels.test(['-ra', '--skip-examples'])
```
Comment (bashtage) [inline tools/ci/run_statsmodels_test_suite.js:46]: Should add --skip-slow too.

Comment (agriyakhetarpal) [reply tools/ci/run_statsmodels_test_suite.js:46]: Sure – let me push this in the next commit. At the current stage of the PR, the test suite with slow tests skipped produces four failures, while without them it produces ten.

**Comment by josef-pkt** [comment]
changes in unit test looks good to me
The only silenced "failure" seems to be in the base tests.

```
    @pytest.mark.xfail(is_wasm(), reason="Fails in WASM, not sure what to do with this yet")
    def test_zero_collinear(self):
```

Question (I never looked at this)
Is the numeric code, numpy, scipy, linear algebra, run in 32 or 64 bit precision?

**Comment by agriyakhetarpal** [comment]
> Question (I never looked at this)
Is the numeric code, numpy, scipy, linear algebra, run in 32 or 64 bit precision?

As far as I understand, because the Python interpreter for Pyodide is 32-bit, and so are the SciPy and NumPy wheels that Pyodide loads from its CDN (i.e., the `wasm32` platform tag), all numeric code runs with 32-bit precision.

I am hoping to dive deeper into the failure above, since this looks like the only one that remains to be fixed: https://github.com/agriyakhetarpal/statsmodels/actions/runs/9404775753/job/25904437721

**Comment by agriyakhetarpal** [comment]
All tests passing at https://github.com/agriyakhetarpal/statsmodels/actions/runs/9506424329/job/26203552820 and skipped/xfailed where applicable: this should be ready for review, now – @bashtage and @rgommers.

I put in some cleanup for the installation instructions for the documentation and removed the `exit=True` command plus some legacy code from the `statsmodels.test()` API to make it work better and to be able to access the return code programmatically (and also not exit the Python interpreter after the test suite proceeds to completion). Please let me know if these ancillary changes are better placed in a follow-up PR or if they are fine here.

As mentioned in the PR description, a follow-up PR would be to push these wheels to Anaconda.org at https://anaconda.org/scientific-python-nightly-wheels/statsmodels, but I guess that's minimal enough to include here itself, please let me know and I can implement that in a subsequent commit :) 

Among other things, I can try to configure [`cibuildwheel` support](https://github.com/pypa/cibuildwheel/pull/1456/) for the workflow as well, but there are some problems with the use of the Node.js-based runner under `CIBW_TEST_COMMAND_PYODIDE` – I have opened an issue about it: https://github.com/pypa/cibuildwheel/issues/1878. The same problem exists for [`scikit-image`](https://github.com/scikit-image/scikit-image/pull/7440) at the time of writing.

File: tools/ci/run_statsmodels_test_suite.js (hunk 1 above)
Comment (agriyakhetarpal) [reply tools/ci/run_statsmodels_test_suite.js]: This was addressed in 59f5f919870290265aaf33065a1a1ce59a72e6a5.

File: tools/ci/run_statsmodels_test_suite.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+const { opendir } = require('node:fs/promises');
+const { loadPyodide } = require("pyodide");
+
+async function main() {
+    let exit_code = 0;
+    try {
+        global.pyodide = await loadPyodide();
+        let pyodide = global.pyodide;
+        const FS = pyodide.FS;
+        const NODEFS = FS.filesystems.NODEFS;
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply tools/ci/run_statsmodels_test_suite.js]: Addressed in 2a47359789d5aee9ffb5b737a17dede9df3a11a5.

File: tools/ci/run_statsmodels_test_suite.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+// A JavaScript file to run the statsmodels test suite using Pyodide
+// This file is used by the GitHub Actions workflow to run the tests
+// against the Pyodide build of statsmodels defined in emscripten.yml.
+
+// The contents of this file are attributed to the scikit-learn developers,
+// who have a similar file in their repository:
+// https://github.com/scikit-learn/scikit-learn/blob/main/build_tools/azure/pytest-pyodide.js
+
+
+const { opendir } = require('node:fs/promises');
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply tools/ci/run_statsmodels_test_suite.js:10]: This is required later in the code, I guess. I can't remove this for now

File: statsmodels/stats/tests/test_rates_poisson.py [hunk 4]
```diff
@@ -716,10 +717,12 @@
         assert_allclose(tst2.pvalue, tst.pvalue, rtol=rtol)
 
         # check corner case count2 = 0, see issue #8313
-        with pytest.warns(RuntimeWarning):
-            tst = smr.test_poisson_2indep(count1, n1, 0, n2, method=meth,
-                                          compare=compare,
-                                          value=None, alternative='two-sided')
+        if not is_wasm():  # No fp exception support in WASM
+            with pytest.warns(RuntimeWarning):
+                tst = smr.test_poisson_2indep(
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/stats/tests/test_rates_poisson.py:722]: Not sure about why this existed as such and this was present before I changed this file through my PR – perhaps we can just call the callable `smr.test_poisson_2indep()` here with its arguments without having to assign it to this variable `tst`?

File: statsmodels/tools/tools.py [hunk 5]
```diff
@@ -18,6 +20,10 @@ def asstr2(s):
         return str(s)
 
 
+def is_wasm():
```
Comment (bashtage) [inline statsmodels/tools/tools.py:23]: this function should be in compat/python.py since it is for specific python version compatability.   I think you should also define it as a constant, `IS_WASM` or similar, which is more the style of the checks in `@skip` or `@xfail` decorators (e.g., `PANDAS_LT_3`).

File: statsmodels/base/tests/test_generic_methods.py [hunk 6]
```diff
@@ -196,7 +198,12 @@ def test_zero_collinear(self):
         # TODO: Can we choose a test case without this issue?
         #  If not, should we be getting this warning for all
         #  model subclasses?
+        # TODO: Investigate how to resolve unseen warnings for Pyodide
+        # Most likely coming from NumPy.linalg + lack of fp exceptions
+        # support under WASM
         warn_cls = HessianInversionWarning if isinstance(mod, sm.GLM) else None
```
Comment (bashtage) [inline statsmodels/base/tests/test_generic_methods.py:204]: ` warn_cls = HessianInversionWarning if isinstance(mod, sm.GLM) and not is_wasm() else None` and remove indentation

**Review by bashtage — CHANGES_REQUESTED** [review]
Some things that can be simplified.

File: statsmodels/stats/tests/test_rates_poisson.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (josef-pkt) [reply statsmodels/stats/tests/test_rates_poisson.py:722]: assigment to tst  can be removed
(The only assert I can think of that will work with all methods is checking continuity of pvalues as in 
https://github.com/statsmodels/statsmodels/issues/8313#issuecomment-1157810004 
)

File: statsmodels/tools/tools.py (hunk 5 above)
Comment (agriyakhetarpal) [reply statsmodels/tools/tools.py]: Addressed in bf68d552db092c33a92df20ac18c4a468e8f381c.

File: statsmodels/tsa/tests/test_stattools.py [hunk 7]
```diff
@@ -350,8 +351,9 @@ def test_yw(self):
         assert_almost_equal(pacfyw[1:], self.pacfyw, DECIMAL_8)
 
     def test_yw_singular(self):
-        with pytest.warns(ValueWarning):
-            pacf(np.ones(30), nlags=6)
+        if not is_wasm():  # No fp exception support in WASM
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/tsa/tests/test_stattools.py]: Addressed in c606e2973cb01b251118128044b312d004755da1.

File: statsmodels/tests/test_package.py [hunk 8]
```diff
@@ -1,7 +1,11 @@
 import subprocess
 import sys
 
+from statsmodels.tools.tools import is_wasm
+import pytest
 
+
+@pytest.mark.skipif(is_wasm(), reason="Can't start subprocess in WASM/Pyodide")
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/tests/test_package.py]: Done, I hope this is completed through bf68d552db092c33a92df20ac18c4a468e8f381c and 88d883cf73507f753bce1467e33af36f4368ad69.

File: statsmodels/stats/tests/test_pairwise.py [hunk 9]
```diff
@@ -245,6 +247,7 @@ def test_table_names_custom_group_order(self):
             assert_((first_group, second_group) == expected_order[i - 1])
 
 
+@pytest.mark.xfail(is_wasm(), reason="Failing on Pyodide due to issues with scipy.optimize's solver")
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/stats/tests/test_pairwise.py]: Thanks, the linter did not run on this file earlier after I had edited and committed the changes. Addressed in 2873bb8e2faaaef3771d58a76e0e5dcfd6a218ba.

File: statsmodels/graphics/tests/test_functional.py [hunk 10]
```diff
@@ -77,6 +82,10 @@ def test_hdr_basic_brute(close_figures, reset_randomstate):
         pytest.xfail('Multiprocess randomly crashes in Windows testing')
 
 
+@pytest.mark.skipif(
+    is_wasm(),
+    reason="Multiprocessing is not supported in WASM/Pyodide"
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/graphics/tests/test_functional.py:83]: Thanks, I have temporarily disabled the `skipif` marker in 5417124904dade237bc58c287d865bb6ff17f676. Let me see if this makes the test fail.

File: statsmodels/base/tests/test_generic_methods.py (hunk 6 above)
Comment (agriyakhetarpal) [reply statsmodels/base/tests/test_generic_methods.py]: Addressed in 1f28d5783f4026742358944f3650fc2d7ecf7f5f.

**Comment by josef-pkt** [comment]
`exit=False` in test runner

The purpose is to keep the interpreter alive after running a test.
The usecase was if we want to check several specific directories in the current python session.

Running the tests is the easiest for users to run unit tests because they don't need to be familiar with pytest commands.
The function was initially written for nosetests, but pytest behaves in the same way AFAICS.

It will be better to keep the current default.

example
```
>>> import statsmodels.othermod as om
>>> om.test()
Running pytest ...\statsmodels_gh\statsmodels\statsmodels\othermod --tb=short --disable-pytest-warnings
================================================= test session starts =================================================
platform win32 -- Python 3.11.4, pytest-7.2.1, pluggy-1.0.0
rootdir: ...\statsmodels_gh\statsmodels, configfile: setup.cfg
plugins: anyio-3.6.2, dash-2.9.3, hypothesis-6.75.2, nbval-0.9.6
collected 13 items

...\statsmodels_gh\statsmodels\statsmodels\othermod\tests\test_beta.py .............      [100%]

================================================= 13 passed in 0.16s ==================================================

>>> om.test(exit=True)
Running pytest ...\statsmodels_gh\statsmodels\statsmodels\othermod --tb=short --disable-pytest-warnings
================================================= test session starts =================================================
platform win32 -- Python 3.11.4, pytest-7.2.1, pluggy-1.0.0
rootdir: ...\statsmodels_gh\statsmodels, configfile: setup.cfg
plugins: anyio-3.6.2, dash-2.9.3, hypothesis-6.75.2, nbval-0.9.6
collected 13 items

...\statsmodels_gh\statsmodels\statsmodels\othermod\tests\test_beta.py .............      [100%]

================================================= 13 passed in 0.16s ==================================================
Exit status: 0

C:\Users\...\WPy64-31140\scripts>
```

**Comment by bashtage** [comment]
One more thought.  Is it possible to spit the pyodide install step from the test step.  Would make it easier to maintain.

File: statsmodels/stats/tests/test_rates_poisson.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/stats/tests/test_rates_poisson.py]: Thanks, addressed in 44ccd6d15c9e7d9703f081b76bdac8002a1b5d27.

File: statsmodels/datasets/tests/test_utils.py [hunk 11]
```diff
@@ -1,18 +1,25 @@
...
 from statsmodels.datasets import get_rdataset, webuse, check_internet, utils
+from statsmodels.tools.tools import is_wasm
 
 cur_dir = os.path.dirname(os.path.abspath(__file__))
 
-IGNORED_EXCEPTIONS = (HTTPError, URLError, SSLError,  UnicodeEncodeError,
-                      timeout)
+if not is_wasm():
+    from ssl import SSLError
+    IGNORED_EXCEPTIONS = (
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/datasets/tests/test_utils.py]: Addressed in 17d388e00c0efc461360cba7fdd5b6623286ac31.

File: statsmodels/compat/python.py [hunk 12]
```diff
@@ -1,10 +1,17 @@
...
 Compatibility tools for differences between Python 2 and 3
 """
+
+import platform
 import sys
 from typing import Literal
 
+
 PY37 = sys.version_info[:2] == (3, 7)
+PYTHON_IMPL_WASM = (
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/compat/python.py:11]: It's probably reasonable to ignore this warning – we're using this constant in many a test file

**Comment by agriyakhetarpal** [comment]
> One more thought. Is it possible to spit the pyodide install step from the test step. Would make it easier to maintain.

Could you elaborate, @bashtage? i.e., here

```yaml
    - name: Set up Pyodide virtual environment and test statsmodels for Pyodide
      run: |
        pyodide venv .venv-pyodide
        source .venv-pyodide/bin/activate
        npm install pyodide@${{ env.PYODIDE_VERSION }}
        node tools/ci/run_statsmodels_test_suite.js
```

we create a Pyodide xbuildenv that relies on Node.js (in the step above it). The `statsmodels` WASM wheel is built in the step above using `pyodide build` which uses a monkeypatched rendition of `pypa/build`, so that's separate from the testing step already.

File: statsmodels/compat/python.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bashtage) [reply statsmodels/compat/python.py:11]: Add it to __all__ and that warning will disappear.

[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/compat/python.py:11]: Addressed in 6d7ad90296197878d22a87eb7d8792a9bca4cbb4.

**Comment by agriyakhetarpal** [comment]
> `exit=False` in test runner
>
> The purpose is to keep the interpreter alive after running a test.
> The usecase was if we want to check several specific directories in the current python session.
>
> Using the test runner function is the easiest way for users to run unit tests because they don't need to be familiar with pytest commands.
> The function was initially written for nosetests, but pytest behaves in the same way AFAICS.
>
> It will be better to keep the current default.

Thanks for your feedback and the additional context, @josef-pkt – I've reverted the change in commits 19e6af8b0bcb96de991f0262e9c677bc4105420e and fe2ee9c489e39cccdaf8afb30f48ba36d09c5e57, but I kept the `return (status == 0)` statement there so that I can assign an integer/boolean value of the test suite results to a variable instead of receiving `None` – I found it useful when trying to exit the tests properly inside the `await pyodide.runPythonAsync` function in the JavaScript file. This way, this should restore the original behaviour but with an additional return value from `__call__` instead of not having one. Please let me know if this change is undesirable and I would be happy to revert it. :)

File: statsmodels/tools/tools.py [hunk 13]
```diff
@@ -18,17 +18,6 @@ def asstr2(s):
...
-def _make_dictnames(tmp_arr, offset=0):
-    """
-    Helper function to create a dictionary mapping a column number
-    to the name in tmp_arr.
-    """
-    col_map = {}
-    for i, col_name in enumerate(tmp_arr):
-        col_map[i + offset] = col_name
-    return col_map
-
```
Comment (agriyakhetarpal) [inline statsmodels/tools/tools.py:30]: This function was removed because I noticed that this function was not being used anywhere, and therefore my code editor's linter (Ruff) raised a warning about this.

File: statsmodels/tools/_testing.py [hunk 14]
```diff
@@ -29,21 +28,22 @@ def __init__(self, package_path=None):
         self.package_path = os.path.dirname(package_path)
         self.package_name = f.f_locals.get('__name__', None)
 
-    def __call__(self, extra_args=None, exit=False):
```
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/tools/_testing.py:32]: Reverted this change based on the suggestions received.

File: statsmodels/graphics/tests/test_functional.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (agriyakhetarpal) [reply statsmodels/graphics/tests/test_functional.py:83]: It looks like the test hasn't failed, so we can resolve this conversation for now. Thanks for the pointer, @bashtage!

File: statsmodels/datasets/tests/test_utils.py [hunk 15]
```diff
@@ -1,18 +1,22 @@
...
 import pytest
 
+from statsmodels.compat.python import PYTHON_IMPL_WASM
 from statsmodels.datasets import get_rdataset, webuse, check_internet, utils
 
-cur_dir = os.path.dirname(os.path.abspath(__file__))
 
-IGNORED_EXCEPTIONS = (HTTPError, URLError, SSLError,  UnicodeEncodeError,
-                      timeout)
+CUR_DIR = os.path.dirname(os.path.abspath(__file__))
```
Comment (agriyakhetarpal) [inline statsmodels/datasets/tests/test_utils.py:14]: `cur_dir` was being used multiple times in this file, so I had marked this as a constant as well in f20d90bb00634e6f1369b1ac655e1ef91936c5ad.

File: .github/workflows/emscripten.yml [hunk 16]
```diff
@@ -0,0 +1,70 @@
+# Attributed to NumPy via https://github.com/numpy/numpy/pull/25894
+# https://github.com/numpy/numpy/blob/d2d2c25fa81b47810f5cbd85ea6485eb3a3ffec3/.github/workflows/emscripten.yml
+name: Test Emscripten/Pyodide build
+
+on:
+  # TODO: remove after ready to merge
+  push:
+    # branches: [ 'main']
+  pull_request:
+    # branches: [ 'main' ]
```
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:10]: Marking this in an explicit comment so that we remember to fix this

File: .github/workflows/emscripten.yml [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  build-wasm-emscripten:
+    name: Build statsmodels distribution for Pyodide
+    runs-on: ubuntu-latest
+    env:
+      PYODIDE_VERSION: 0.26.0
```
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:23]:
```suggestion
      PYODIDE_VERSION: 0.26.1
```

Pyodide version 0.26.1 is out now

**Comment by agriyakhetarpal** [comment]
Thanks for the reviews, @bashtage and @josef-pkt! I have resolved all of the conversations but a few – this should be ready for another round of review. Based on https://github.com/statsmodels/statsmodels/pull/9270#issuecomment-2166807830, two things remain – pushing these wheels to Anaconda.org and `cibuildwheel`, out of which the latter is stalled. Please let me know if it would be okay to configure the former here.

Another thing I just noticed at the time of writing this comment is that https://github.com/agriyakhetarpal/statsmodels/actions/runs/9520858185/job/26247012073 was marked as passing even if a test failed. I'll make changes for that and fix the failing test.

**Comment by agriyakhetarpal** [comment]
Tests are passing here: https://github.com/agriyakhetarpal/statsmodels/actions/runs/9522416223/job/26251913042, but I need to adjust the script to ensure that it fails with the right exit code and that it doesn't fail even if the tests pass.

**Comment by josef-pkt** [comment]
> but I kept the return (status == 0) statement 

sounds fine to me, I don't see a problem with an extra returned value

**Comment by agriyakhetarpal** [comment]
> I need to adjust the script to ensure that it fails with the right exit code and that it doesn't fail even if the tests pass.

Took me a bit longer than expected, but the test suite isn't acting up anymore: https://github.com/agriyakhetarpal/statsmodels/actions/runs/9618655879/job/26533005508 passed, and https://github.com/agriyakhetarpal/statsmodels/actions/runs/9618990442/job/26534075665 failed like I wanted it to – as expected. Ready for review, again!

File: .github/workflows/emscripten.yml (hunk 16 above)
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Addressed in adbf35326eb0e3e9c0f2f5104d68a8a7d7d12c56.

**Comment by agriyakhetarpal** [comment]
Merged the `main` branch to get the latest changes – I think that the Azure Pipelines failures are unrelated: `ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject`, but I see that `statsmodels` since version 0.14.2 already has NumPy v2 support. This is coming from another dependency, perhaps?

**Comment by agriyakhetarpal** [comment]
Hi, @bashtage and @josef-pkt! Is it okay to tag you both to seek another review on this PR, since it has been a month since my last review request? It is complete and ready to go from my end, and I've merged the latest changes from `main` now and resolved the conflicts that came up – thanks! I'm glad to make this PR easier to review if needed, so if anything is required to be done, please let me know and I'll take a look at that.

**Comment by bashtage** [comment]
Thanks for the work.  I'll get around to it next week.

**Comment by agriyakhetarpal** [comment]
Sure, thanks, @bashtage! Could you trigger yhe workflows in the meantime here, please?

**Comment by agriyakhetarpal** [comment]
Hi again, @bashtage – here is a gentle reminder since you mentioned you would get around to it the week after. There is no hurry here whatsoever – whenever you get the time. Thanks!

**Comment by bashtage** [comment]
I am away this week but will absolutely take a look when I'm back.

File: statsmodels/graphics/tests/test_gofplots.py [hunk 18]
```diff
@@ -98,6 +103,10 @@ def test_ppplot_other_prbplt(self, close_figures):
 
     @pytest.mark.xfail(strict=True)
     @pytest.mark.matplotlib
+    @pytest.mark.skipif(
+        PYTHON_IMPL_WASM,
+        reason="Matplotlib uses different backend in WASM/Pyodide"
```
Comment (bashtage) [inline statsmodels/graphics/tests/test_gofplots.py:108]: Does the backend matter for these and the others that are marked to skip?

File: statsmodels/compat/python.py [hunk 19]
```diff
@@ -19,6 +26,8 @@
     "lrange",
     "lfilter",
     "with_metaclass",
+    PY37,
```
Comment (bashtage) [inline statsmodels/compat/python.py:29]: Should these be surrounded by `""`

Comment (bashtage) [reply statsmodels/compat/python.py:29]: I don't think we support PY37 anymore officially, could remove

**Review by bashtage — CHANGES_REQUESTED** [review]
One small change and one quesation.  Overall LGTM. I can't really comment n the wasm/js stuff.

File: statsmodels/graphics/tests/test_gofplots.py (hunk 18 above)
Comment (agriyakhetarpal) [reply statsmodels/graphics/tests/test_gofplots.py:108]: If you mean the functionality of the plots, I don't think so. Normally, one will not need to adapt the code to use `matplotlib-pyodide` to use the HTML5/WASM canvases, since it gets installed when one installs `matplotlib` using `micropip`. The details are a little murky since it has been some time for me, but I received the error: `'Line2D' object has no attribute "other"`, so it is probably a missing feature in `matplotlib-pyodide` at the time of writing.

For example, https://github.com/statsmodels/statsmodels/blob/main/examples/python/categorical_interaction_plot.py (I can't find any other failing example right now) can be tried out in the Pyodide console (https://pyodide.org/en/stable/console.html), though it won't display any of the `matplotlib.figure.Figure` objects, of course, lacking a canvas for the renderer – they can be viewed in a JupyterLite notebook (https://jupyter.org/try-jupyter/lab/) and they should all work without any code changes. Not all of the tests marked with `@pytest.mark.matplotlib` have been skipped, and the ones skipped here are those that were failing because of the error noted above. I could investigate further as a follow-up if you would want me to do so.

File: statsmodels/compat/python.py (hunk 19 above)
Comment (agriyakhetarpal) [reply statsmodels/compat/python.py]: Good plan – addressed in 8ed424e02717fb1263eee2973c6c57c1ba2a50fd and b6353804fd71c8d1adc60b1dd83197d974e6fb0b.

Comment (agriyakhetarpal) [reply statsmodels/compat/python.py]: and I wrapped `PYTHON_IMPL_WASM` in `__all__` in quotes in 98005e1f55a561bb3ace51a0b0f251f55d25b71b.

**Comment by bashtage** [comment]
Going to merge this and we'll see if other changes are needed.

**Comment by bashtage** [comment]
I'll let it run one more time to make sure there is nothing stange

**Review by bashtage — APPROVED** [review]
Approve
